java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -133,30 +135,71 @@ public void scan(Iterable<InputFile> sourceFiles, Iterable<InputFile> testFiles,
+   * This batch size is then used as a threshold: files are added to a batch until the threshold is passed.
+   * Once the threshold is passed, the batch is processed for analysis.
+   *
+   * If no batch size is configured, the input files are scanned as a single batch.
+   *
+   * @param inputFiles The collections of files to scan
+   */
+  private void scanAsBatch(Iterable<? extends InputFile>... inputFiles) {
+    List<InputFile> files = new ArrayList<>();
+    for (Iterable<? extends InputFile> group: inputFiles) {
+      files.addAll(astScanner.filterModuleInfo(group));
+    }
+
+    // Compute the batch size
+    long minBatchModeSize = MAX_BATCH_SIZE;
+    long batchModeSizeInMB = sonarComponents.getBatchModeSizeInMB();
+    if (batchModeSizeInMB != -1L) {
+      minBatchModeSize = batchModeSizeInMB * 1_000_000L;
+      // Bound the memory usage per batch to a maximum of 10% of total memory
+      minBatchModeSize = Math.min(minBatchModeSize, MAX_BATCH_SIZE);
+    }
+    LOG.debug("Scanning with batch size {} B", minBatchModeSize);
+
     try {
-      List<InputFile> allFiles = new ArrayList<>();
-      Arrays.stream(sourceFiles).forEach(files -> files.forEach(allFiles::add));
-      try {
-        JParserConfig.Mode.BATCH
-          .create(JParserConfig.effectiveJavaVersion(javaVersion), globalClasspath)
-          .parse(allFiles, this::analysisCancelled, this::scanAsBatchCallback);
-      } finally {
-        astScanner.endOfAnalysis();
-        astScannerForTests.endOfAnalysis();
-        astScannerForGeneratedFiles.endOfAnalysis();
+      int batchFirstPos = 0;
+      while (batchFirstPos < files.size()) {
+        long batchSize = files.get(batchFirstPos).contents().length();
+        int batchLastPosExclusive = batchFirstPos + 1;
+        while (batchSize < minBatchModeSize && batchLastPosExclusive < files.size()) {
+          batchSize += files.get(batchLastPosExclusive).contents().length();
+          batchLastPosExclusive++;
+        }
+        List<InputFile> batch = files.subList(batchFirstPos, batchLastPosExclusive);
+        scanBatch(batch);
+        batchFirstPos = batchLastPosExclusive;
       }
     } catch (AnalysisException e) {
       throw e;
     } catch (Exception e) {
       astScanner.checkInterrupted(e);
-      LOG.error("Batch Mode failed, analysis of Java Files stopped.", e);
+      astScannerForTests.checkInterrupted(e);
+      astScannerForGeneratedFiles.checkInterrupted(e);
+      LOG.error(BATCH_ERROR_MESSAGE, e);
       if (astScanner.shouldFailAnalysis()) {
-        throw new AnalysisException("Batch Mode failed, analysis of Java Files stopped.", e);
+        throw new AnalysisException(BATCH_ERROR_MESSAGE, e);
       }
     }
   }
 
+  private<T extends InputFile> void scanBatch(List<T> allFiles) {
+    try {
+      JParserConfig.Mode.BATCH
+        .create(JParserConfig.effectiveJavaVersion(javaVersion), globalClasspath)
+        .parse(allFiles, this::analysisCancelled, this::scanAsBatchCallback);
```

Comment:
In the `parse` method, we log the following block for each batch, we could have a lot of:
```
[INFO] Using ECJ batch to parse source files.
[INFO] Starting batch processing.
[INFO] 100% analyzed
[INFO] Batch processing: Done!
```